File: ruby/lib/br_validator/ie/ap.rb
# frozen_string_literal: true

require_relative "../shared"
require_relative "mod11"

module BrValidator
  module Ie
    # AP: verified against the official SEFAZ-AP "Roteiro de Crítica da
    # Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_AP.html),
    # including its worked example (030123459).
    #
    # Format: "03" (fixed) + 6 sequence digits + 1 check digit = 9 digits
    # total. Unlike other states, the weighted sum starts from a constant
    # "p" that depends on the numeric range of the registration, and a
    # zero remainder maps to a range-dependent digit "d" instead of always
    # 0.
    module Ap
      module_function

      ALLOWED_CHARS = /\A[\d\s]+\z/.freeze
      WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2].freeze
      private_constant :ALLOWED_CHARS, :WEIGHTS

      def normalize(value)
        Shared.remove_non_digits(value)
      end

      def resolve_constants(base)
        return [5, 0] if base <= 3_017_000
        return [9, 1] if base <= 3_019_022

        [0, 0]
      end
      private_class_method :resolve_constants

      def calculate_check_digit(v)
        base = v[0...8]
        p, d = resolve_constants(base.to_i)
        sum = p + Mod11.weighted_sum(base, WEIGHTS)
        remainder = sum % 11

        return 0 if remainder == 1
        return d if remainder.zero?

        11 - remainder
      end
      private_class_method :calculate_check_digit

      def is_valid?(value)
        return false unless value.match?(ALLOWED_CHARS)

        v = normalize(value)

        return false unless v.length == 9 && v.start_with?("03")

        calculate_check_digit(v) == v[8].to_i
      end

      def format(value)
        normalize(value)
      end
    end
  end
end

File: python/src/br_validator/ie/_ap.py
"""AP: verified against the official SEFAZ-AP "Roteiro de Crítica da
Inscrição Estadual" (sintegra.gov.br/Cad_Estados/cad_AP.html), including
its worked example (030123459).

Format: "03" (fixed) + 6 sequence digits + 1 check digit = 9 digits total.
Unlike other states, the weighted sum starts from a constant "p" that
depends on the numeric range of the registration, and a zero remainder
maps to a range-dependent digit "d" instead of always 0.
"""

import re

from .._shared import remove_non_digits
from ._mod11 import weighted_sum

_ALLOWED_CHARS = re.compile(r"^[\d\s]+$")
_WEIGHTS = [9, 8, 7, 6, 5, 4, 3, 2]


def normalize(value: str) -> str:
    return remove_non_digits(value)


def _resolve_constants(base: int):
    if base <= 3017000:
        return 5, 0

    if base <= 3019022:
        return 9, 1

    return 0, 0


def _calculate_check_digit(v: str) -> int:
    base = v[:8]
    p, d = _resolve_constants(int(base))
    total = p + weighted_sum(base, _WEIGHTS)
    remainder = total % 11

    if remainder == 1:
        return 0

    if remainder == 0:
        return d

    return 11 - remainder


def is_valid(value: str) -> bool:
    if not _ALLOWED_CHARS.match(value):
        return False

    v = normalize(value)

    if len(v) != 9 or not v.startswith("03"):
        return False

    return _calculate_check_digit(v) == int(v[8])


def format(value: str) -> str:
    return normalize(value)
